designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,188 @@
+# Disruption Controls By Action 
+## User Scenarios 
+1. I as a user would like to ensure upgrades only occur during business hours or stricter windows. I would also like to not restrict the cost savings of consolidation when blocking upgrade via drift.
```

Comment:
Trying to lay out what is being implicitly being said here. I haven't ran node disruption budget yet in my cluster so its really hard for me to figure out how useful bucketing my disruption budget is however I do see value in allowing for reprioritization based on a schedule.

The need for bucketing means that during a disruption evaluation window a lot of nodes are always contenting for disruption but because of the existing order and singular budget the desired disruption action isn't taking place. Not only that, one or some of the disruption action is so dominate in the cluster that it completely drains the budget preventing the desired disruption to take place. Bucketing allows mixing of disruption depending on a schedule.

I think its a interesting use case but would love to know if this is a problem users faces. I think being able to just choose or reorder disruption action could be enough for our us.

Reply:
>  I think being able to just choose or reorder disruption action could be enough for our us.


Lets say i like having upgrade go before consolidation during the weekend, but I want to allow for consolidation over upgrade during the week. Im not sure how with the second design we can respect that ordering. With the first design we can define an actionWeight to a budget if we define the action alongside the budget. 

I think we only really have a need or want to add a actionWeight if the `nodes` field were to be a global field and not a per action field.  If we can find a way to easily reason about the node amounts we set in the budgets alongside actions, this actionWeight wouldn't need to exist. 

@njtran  do you know if there are any assumptions in the disruption controller that make assumptions based on the order the actions are called?

Reply:
I decided against changing the ordering in the disruptino controller all together, read more in the Clarifying Behavior section